### PR TITLE
[Performance] BEVFormer: Build SCA rebatch plan once per frame

### LINKED ISSUE #55193: [BEVFormer] Build the SCA rebatch plan once per frame
## Context

All encoder layers process the same camera visibility mask and projected reference points for a frame. The corresponding SCA indices, lengths, and reference-point rebatch are derived inside each layer call.

## Problem

- [bev_mask readback, max_len, camera indices, and helper tensors are rebuilt inside each SCA call](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_spatial_cross_attention.py#L152-L170), six times per frame.
- [Reference-point rebatching is repeated inside each layer](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_spatial_cross_attention.py#L194-L230) even though it does not depend on the layer's query.
- Repeating the plan adds host work and prevents the encoder loop from becoming a clean device-execution region.

## Proposed direction

- Introduce a frame-level rebatch plan [immediately after point sampling and before the layer loop](https://github.com/tenstorrent/tt-metal/blob/main/models/experimental/bevformer/tt/tt_encoder.py#L464-L497).
- Store camera query IDs, common lengths, scatter indices, contributor counts, and rebatched reference points in the plan.
- Reuse the plan in every layer; keep only the query gather and attention result accumulation layer-dependent.

## Acceptance criteria

- [x] Layer-invariant planning and reference-point rebatching run once per encoder forward.
- [x] Every layer consumes the same plan without changing SCA semantics.
- [x] Existing SCA, layer, and encoder PCC suites pass at their configured thresholds, including PCC >= 0.997 for the base encoder.
- [x] Encoder-level profiling demonstrates reduced repeated host work using the same input and configuration.

## References

- [Prototype implementation](https://github.com/tenstorrent/tt-metal/commit/d10beef5a04f7362351b76e73781ef5bdee42a18)

## Dependencies

- Depends on #55192

### PR DESCRIPTION
**Pull request by ctr-mmicicTT**
### PR Category
Performance

### Summary
Build the SCA rebatch plan once per encoder forward, immediately after point sampling. Camera visibility, indices, scatter IDs, contributor counts, and rebatched reference points are frame-invariant; each layer now only gathers query and accumulates attention. Standalone SCA still builds the plan internally.

Closes #55193
Depends on #55494

### Performance

Single encoder layer, Blackhole P100.
Source: `models/experimental/bevformer/tests/perf/test_layer_perf.py::test_bevformer_layer_perf`

| Metric | Before ([a473dd3](https://github.com/tenstorrent/tt-metal/commit/a473dd32727abb917925b95ce9cd6b6357a71cf1)) | After ([5b1b911](https://github.com/tenstorrent/tt-metal/commit/5b1b9115dd8d6cb0fdf4261c8a0c1ecd298ec511)) | Delta |
| --- | --- | --- | --- |
| Measured forward elapsed | 324.126 ms | 319.0 ms | -5.1 ms (-1.6%) |
| Op-to-op gaps | 4.670 ms | 0.872 ms | **-3.80 ms (-81%)** |
| Summed device kernels | 319.456 ms | 318.128 ms | -1.3 ms (within noise) |

The gain is the removal of the per-layer host planning gap: op-to-op gap time drops by 81%, while summed device kernel time is unchanged. Full reports and screenshots: #55193 ([before](https://github.com/tenstorrent/tt-metal/issues/55193#issuecomment-5665384321), [after](https://github.com/tenstorrent/tt-metal/issues/55193#issuecomment-5665683301)). Raw `tt-perf-report` CSVs for both runs are attached below:
[tt_perf_report.csv before](https://github.com/user-attachments/files/32244704/tt_perf_report.csv)
[tt_perf_report.csv after](https://github.com/user-attachments/files/32244751/tt_perf_report.csv)

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ctr-mmicic/55193-bevformer-sca-rebatch-plan-once-per-frame)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ctr-mmicic/55193-bevformer-sca-rebatch-plan-once-per-frame)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ctr-mmicic/55193-bevformer-sca-rebatch-plan-once-per-frame)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ctr-mmicic/55193-bevformer-sca-rebatch-plan-once-per-frame)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ctr-mmicic/55193-bevformer-sca-rebatch-plan-once-per-frame)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ctr-mmicic/55193-bevformer-sca-rebatch-plan-once-per-frame)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ctr-mmicic/55193-bevformer-sca-rebatch-plan-once-per-frame)

### CODE REVIEW

**Comment by ctr-mmicicTT** [comment]
CI tests: https://github.com/tenstorrent/tt-metal/actions/runs/34837154288

File: models/experimental/bevformer/tt/tt_encoder.py [hunk 1]
```diff
@@ -469,6 +472,13 @@ def forward(
                 device=self.device,
             )
 
+        # Camera projection is per-frame, so the SCA rebatch plan is too. Layers only gather query.
+        rebatch_plan = (
+            build_rebatch_plan(reference_points_cam, bev_mask, self.embed_dims, self.device)
```
[the remark this answers is not in the document]
Comment (ctr-mmicicTT) [reply models/experimental/bevformer/tt/tt_encoder.py:477]: Before/after results are present in https://github.com/tenstorrent/tt-metal/issues/55193 comment section.

[the remark this answers is not in the document]
Comment (ctr-mmicicTT) [reply models/experimental/bevformer/tt/tt_encoder.py:477]: Will be addressed in https://github.com/tenstorrent/tt-metal/issues/55190

File: models/experimental/bevformer/tt/tt_spatial_cross_attention.py [hunk 2]
```diff
@@ -172,87 +299,32 @@ def forward(
...
-        # the contributor count stay on the host; the gathers and the scatter run on device.
-        bev_mask_torch = ttnn.to_torch(bev_mask)
-
-        valid_per_cam = bev_mask_torch.sum(-1) > 0  # [num_cams, B, num_queries]
-        max_len = int(valid_per_cam.sum(-1).max().item())
-
-        if ENABLE_LOGGING:
-            logger.info(f"SCA Valid Queries: {valid_per_cam.sum(-1).flatten().tolist()}")
+        # Every encoder layer in a forward shares one plan; building it here is the standalone path.
+        if rebatch_plan is None:
```
[the remark this answers is not in the document]
Comment (ctr-mmicicTT) [reply models/experimental/bevformer/tt/tt_spatial_cross_attention.py:303]: Added in https://github.com/tenstorrent/tt-metal/commit/d350a2dcf8d1f87c76bf6684c455675a7c1c0bd9

**Comment by ctr-mmicicTT** [comment]
CI test that passed: https://github.com/tenstorrent/tt-metal/actions/runs/34837154288

File: models/experimental/bevformer/tt/tt_encoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (njovanovicTT) [reply models/experimental/bevformer/tt/tt_encoder.py:477]: Please also add the summarized version in the description of the PR. Include perf report `csv` files.

[the remark this answers is not in the document]
Comment (ctr-mmicicTT) [reply models/experimental/bevformer/tt/tt_encoder.py:477]: Summary and CSVs added to the PR description.

**Review by njovanovicTT — APPROVED** [review]
Will approve once the comments are addressed and
